### ISSUE TITLE
Strict RLP decoding / Enforce PaymentTransaction according to spec

### ISSUE THREAD

**Issue by thec00n** [omisego/elixir-omg#1176]
### Description
Recent efforts to achieve a 1:1 mapping from an RLP tx to a decoded PaymentTransaction have recently been merged into the root chain contracts ( PR https://github.com/omisego/plasma-contracts/pull/476). This issue proposes to implement the same mechanisms to achieve a 1:1 mapping and to decode/encode only well-formed PaymentTransactions. 

### Enforce non overlapping RLP decoding
Achieving a 1:1 relationship between decoded and encoded RLP message is not trivial and there are some edge cases that can easily be overlooked. Check out the test inputs in https://github.com/omisego/plasma-contracts/blob/182908d4c7ef669e426356c0bce528a5ebe75593/plasma_framework/test/src/utils/fixture/ and the RLP related test runners in https://github.com/omisego/plasma-contracts/blob/182908d4c7ef669e426356c0bce528a5ebe75593/plasma_framework/test/src/utils/. The implementation in Elixir should cover all of the tests that have been specific on the plasma-contracts.

#### Define Data types 

The current RLPReader implements RLP conform decoding as specified here https://github.com/ethereum/wiki/wiki/RLP.  It also introduces 3 data types that are application specific. 

- [x] RLPReader.toUint():  treats leading zeros as invalid. Must be between 1 and 32 bytes. 
- [x] RLPReader.toAddress() requires the item data to be exactly 20 bytes.
- [x] RLPReader.toBytes32() requires the item data to be exactly 32 bytes.

#### Decode/Encode only valid PaymentTransaction 

During PaymentTransaction decoding it is also checked that the tx is well formed and according to spec: 

- [x]  Transaction inputs must be decoded as bytes32
- [x]  Transaction inputs cannot be empty (all zeros)
- [x]  Transaction outputs cannot be empty (a list that has zero items)
- [x]  Transaction output.amount must be greater than zero
- [x]  Transaction output.outputType must be greater than zero
- [x]  Transaction metadata must be bytes32

CC @kevsul feel free to add details.

**Comment by thec00n** [comment]
Additionally update https://github.com/omisego/elixir-omg/blob/master/docs/transaction_validation.md if changes are required.

**Comment by pdobacz** [comment]
Thanks for the detailed write up, this is very helpful.

I'm currently on top of this, as part of the https://github.com/omisego/security-issues/issues/2 effort and #1164.

It should be relatively straightforward to handle the invalid values and test that on the `elixir-omg` side.

**Comment by pdobacz** [comment]
blocked on https://github.com/poanetwork/ex_abi/issues/22
